

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Case Study 1 - Comparing Tables &mdash; petl v0.8 documentation</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.8',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="petl v0.8 documentation" href="index.html" />
    <link rel="next" title="Related Work" href="related_work.html" />
    <link rel="prev" title="petl - Extract, Transform and Load (Tables of Data)" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="related_work.html" title="Related Work"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="petl - Extract, Transform and Load (Tables of Data)"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">petl v0.8 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="case-study-1-comparing-tables">
<h1>Case Study 1 - Comparing Tables<a class="headerlink" href="#case-study-1-comparing-tables" title="Permalink to this headline">¶</a></h1>
<p>This case study illustrates some of the <a class="reference internal" href="index.html#module-petl" title="petl"><tt class="xref py py-mod docutils literal"><span class="pre">petl</span></tt></a> functions
available for doing some simple profiling and comparison of data from
two tables.</p>
<iframe width="853" height="480" src="http://www.youtube.com/embed/Y0FleNEcO9I" frameborder="0" allowfullscreen></iframe><div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>The files used in this case study can be downloaded from the following
link:</p>
<ul class="simple">
<li><a class="reference external" href="http://aliman.s3.amazonaws.com/petl/petl-case-study-1-files.zip">http://aliman.s3.amazonaws.com/petl/petl-case-study-1-files.zip</a></li>
</ul>
<p>Download and unzip the files:</p>
<div class="highlight-python"><pre>$ wget http://aliman.s3.amazonaws.com/petl/petl-case-study-1-files.zip
$ unzip petl-case-study-1-files.zip</pre>
</div>
<p>The first file (<cite>snpdata.csv</cite>) contains a list of locations in the
genome of the malaria parasite <cite>P. falciparum</cite>, along with some basic
data about genetic variations found at those locations.</p>
<p>The second file (<cite>popdata.csv</cite>) is supposed to contain the same list
of genome locations, along with some additional data such as allele
frequencies in different populations.</p>
<p>The main point for this case study is that the first file
(<cite>snpdata.csv</cite>) contains the canonical list of genome locations, and
the second file (<cite>popdata.csv</cite>) contains some additional data about
the same genome locations and therefore should be consistent with the
first file. We want to check whether this second file is in fact
consistent with the first file.</p>
</div>
<div class="section" id="preparing-the-data">
<h2>Preparing the data<a class="headerlink" href="#preparing-the-data" title="Permalink to this headline">¶</a></h2>
<p>Start the Python interpreter and import <a class="reference internal" href="index.html#module-petl" title="petl"><tt class="xref py py-mod docutils literal"><span class="pre">petl</span></tt></a> functions:</p>
<div class="highlight-python"><pre>$ python
Python 2.7.2+ (default, Oct  4 2011, 20:03:08)
[GCC 4.6.1] on linux2
Type "help", "copyright", "credits" or "license" for more information.
&gt;&gt;&gt; from petl import *</pre>
</div>
<p>To save some typing, let <cite>a</cite> be the table of data extracted from the
first file (<cite>snpdata.csv</cite>), and let <cite>b</cite> be the table of data extracted
from the second file (<cite>popdata.csv</cite>), using the <a class="reference internal" href="index.html#petl.fromcsv" title="petl.fromcsv"><tt class="xref py py-func docutils literal"><span class="pre">fromcsv()</span></tt></a>
function:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">fromcsv</span><span class="p">(</span><span class="s">&#39;snpdata.csv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">fromcsv</span><span class="p">(</span><span class="s">&#39;popdata.csv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Examine the header from each file using the <a class="reference internal" href="index.html#petl.header" title="petl.header"><tt class="xref py py-func docutils literal"><span class="pre">header()</span></tt></a>
function:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">header</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">(&#39;Chr&#39;, &#39;Pos&#39;, &#39;Ref&#39;, &#39;Nref&#39;, &#39;Der&#39;, &#39;Mut&#39;, &#39;isTypable&#39;, &#39;GeneId&#39;, &#39;GeneAlias&#39;, &#39;GeneDescr&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">header</span><span class="p">(</span><span class="n">b</span><span class="p">)</span>
<span class="go">(&#39;Chromosome&#39;, &#39;Coordinates&#39;, &#39;Ref. Allele&#39;, &#39;Non-Ref. Allele&#39;, &#39;Outgroup Allele&#39;, &#39;Ancestral Allele&#39;, &#39;Derived Allele&#39;, &#39;Ref. Aminoacid&#39;, &#39;Non-Ref. Aminoacid&#39;, &#39;Private Allele&#39;, &#39;Private population&#39;, &#39;maf AFR&#39;, &#39;maf PNG&#39;, &#39;maf SEA&#39;, &#39;daf AFR&#39;, &#39;daf PNG&#39;, &#39;daf SEA&#39;, &#39;nraf AFR&#39;, &#39;nraf PNG&#39;, &#39;nraf SEA&#39;, &#39;Mutation type&#39;, &#39;Gene&#39;, &#39;Gene Aliases&#39;, &#39;Gene Description&#39;, &#39;Gene Information&#39;)</span>
</pre></div>
</div>
<p>There is a common set of 9 fields that is present in both tables, and
we would like focus on comparing these common fields, however
different field names have been used in the two files. To simplify
comparison, use <a class="reference internal" href="index.html#petl.rename" title="petl.rename"><tt class="xref py py-func docutils literal"><span class="pre">rename()</span></tt></a> to rename some fields in the
second file:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">b_renamed</span> <span class="o">=</span> <span class="n">rename</span><span class="p">(</span><span class="n">b</span><span class="p">,</span> <span class="p">{</span><span class="s">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Coordinates&#39;</span><span class="p">:</span> <span class="s">&#39;Pos&#39;</span><span class="p">,</span> <span class="s">&#39;Ref. Allele&#39;</span><span class="p">:</span> <span class="s">&#39;Ref&#39;</span><span class="p">,</span> <span class="s">&#39;Non-Ref. Allele&#39;</span><span class="p">:</span> <span class="s">&#39;Nref&#39;</span><span class="p">,</span> <span class="s">&#39;Derived Allele&#39;</span><span class="p">:</span> <span class="s">&#39;Der&#39;</span><span class="p">,</span> <span class="s">&#39;Mutation type&#39;</span><span class="p">:</span> <span class="s">&#39;Mut&#39;</span><span class="p">,</span> <span class="s">&#39;Gene&#39;</span><span class="p">:</span> <span class="s">&#39;GeneId&#39;</span><span class="p">,</span> <span class="s">&#39;Gene Aliases&#39;</span><span class="p">:</span> <span class="s">&#39;GeneAlias&#39;</span><span class="p">,</span> <span class="s">&#39;Gene Description&#39;</span><span class="p">:</span> <span class="s">&#39;GeneDescr&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">header</span><span class="p">(</span><span class="n">b_renamed</span><span class="p">)</span>
<span class="go">(&#39;Chr&#39;, &#39;Pos&#39;, &#39;Ref&#39;, &#39;Nref&#39;, &#39;Outgroup Allele&#39;, &#39;Ancestral Allele&#39;, &#39;Der&#39;, &#39;Ref. Aminoacid&#39;, &#39;Non-Ref. Aminoacid&#39;, &#39;Private Allele&#39;, &#39;Private population&#39;, &#39;maf AFR&#39;, &#39;maf PNG&#39;, &#39;maf SEA&#39;, &#39;daf AFR&#39;, &#39;daf PNG&#39;, &#39;daf SEA&#39;, &#39;nraf AFR&#39;, &#39;nraf PNG&#39;, &#39;nraf SEA&#39;, &#39;Mut&#39;, &#39;GeneId&#39;, &#39;GeneAlias&#39;, &#39;GeneDescr&#39;, &#39;Gene Information&#39;)</span>
</pre></div>
</div>
<p>Use <a class="reference internal" href="index.html#petl.cut" title="petl.cut"><tt class="xref py py-func docutils literal"><span class="pre">cut()</span></tt></a> to extract only the fields we&#8217;re interested in
from both tables:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">common_fields</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">,</span> <span class="s">&#39;Ref&#39;</span><span class="p">,</span> <span class="s">&#39;Nref&#39;</span><span class="p">,</span> <span class="s">&#39;Der&#39;</span><span class="p">,</span> <span class="s">&#39;Mut&#39;</span><span class="p">,</span> <span class="s">&#39;GeneId&#39;</span><span class="p">,</span> <span class="s">&#39;GeneAlias&#39;</span><span class="p">,</span> <span class="s">&#39;GeneDescr&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_common</span> <span class="o">=</span> <span class="n">cut</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">common_fields</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b_common</span> <span class="o">=</span> <span class="n">cut</span><span class="p">(</span><span class="n">b_renamed</span><span class="p">,</span> <span class="n">common_fields</span><span class="p">)</span>
</pre></div>
</div>
<p>Inspect the data:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">a_common</span><span class="p">)</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;Chr&#39;  | &#39;Pos&#39;   | &#39;Ref&#39; | &#39;Nref&#39; | &#39;Der&#39; | &#39;Mut&#39; | &#39;GeneId&#39;   | &#39;GeneAlias&#39; | &#39;GeneDescr&#39;                                        |</span>
<span class="go">+========+=========+=======+========+=======+=======+============+=============+====================================================+</span>
<span class="go">| &#39;MAL1&#39; | &#39;91099&#39; | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10&#39; | &#39;rifin&#39;                                            |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;91104&#39; | &#39;A&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10&#39; | &#39;rifin&#39;                                            |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93363&#39; | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93382&#39; | &#39;T&#39;   | &#39;G&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93384&#39; | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93390&#39; | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93439&#39; | &#39;T&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93457&#39; | &#39;C&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;94008&#39; | &#39;T&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;94035&#39; | &#39;C&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">b_common</span><span class="p">)</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;Chr&#39;  | &#39;Pos&#39;   | &#39;Ref&#39; | &#39;Nref&#39; | &#39;Der&#39; | &#39;Mut&#39; | &#39;GeneId&#39;   | &#39;GeneAlias&#39;     | &#39;GeneDescr&#39;                                              |</span>
<span class="go">+========+=========+=======+========+=======+=======+============+=================+==========================================================+</span>
<span class="go">| &#39;MAL1&#39; | &#39;91099&#39; | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;SYN&#39; | &#39;PFA0095c&#39; | &#39;MAL1P1.10,RIF&#39; | &#39;rifin&#39;                                                  |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;91104&#39; | &#39;A&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;NON&#39; | &#39;PFA0095c&#39; | &#39;MAL1P1.10,RIF&#39; | &#39;rifin&#39;                                                  |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93363&#39; | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;NON&#39; | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93382&#39; | &#39;T&#39;   | &#39;G&#39;    | &#39;-&#39;   | &#39;NON&#39; | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93384&#39; | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;NON&#39; | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93390&#39; | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;NON&#39; | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93439&#39; | &#39;T&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;SYN&#39; | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;93457&#39; | &#39;C&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;SYN&#39; | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;94008&#39; | &#39;T&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;NON&#39; | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | &#39;94035&#39; | &#39;C&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;NON&#39; | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+---------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="index.html#petl.fromcsv" title="petl.fromcsv"><tt class="xref py py-func docutils literal"><span class="pre">fromcsv()</span></tt></a> function does not attempt to parse any of the
values from the underlying CSV file, so all values are reported as
strings. However, the &#8216;Pos&#8217; field should be interpreted as an integer.</p>
<p>Also, the &#8216;Mut&#8217; field has a different representation in the two
tables, which needs to be converted before the data can be compared.</p>
<p>Use the <a class="reference internal" href="index.html#petl.convert" title="petl.convert"><tt class="xref py py-func docutils literal"><span class="pre">convert()</span></tt></a> function to convert the type of the &#8216;Pos&#8217;
field in both tables and the representation of the &#8216;Mut&#8217; field in
table <cite>b</cite>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">a_conv</span> <span class="o">=</span> <span class="n">convert</span><span class="p">(</span><span class="n">a_common</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b_conv1</span> <span class="o">=</span> <span class="n">convert</span><span class="p">(</span><span class="n">b_common</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b_conv2</span> <span class="o">=</span> <span class="n">convert</span><span class="p">(</span><span class="n">b_conv1</span><span class="p">,</span> <span class="s">&#39;Mut&#39;</span><span class="p">,</span> <span class="p">{</span><span class="s">&#39;SYN&#39;</span><span class="p">:</span> <span class="s">&#39;S&#39;</span><span class="p">,</span> <span class="s">&#39;NON&#39;</span><span class="p">:</span> <span class="s">&#39;N&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b_conv</span> <span class="o">=</span> <span class="n">b_conv2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">a_conv</span><span class="p">)</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;Chr&#39;  | &#39;Pos&#39; | &#39;Ref&#39; | &#39;Nref&#39; | &#39;Der&#39; | &#39;Mut&#39; | &#39;GeneId&#39;   | &#39;GeneAlias&#39; | &#39;GeneDescr&#39;                                        |</span>
<span class="go">+========+=======+=======+========+=======+=======+============+=============+====================================================+</span>
<span class="go">| &#39;MAL1&#39; | 91099 | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10&#39; | &#39;rifin&#39;                                            |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 91104 | &#39;A&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10&#39; | &#39;rifin&#39;                                            |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93363 | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93382 | &#39;T&#39;   | &#39;G&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93384 | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93390 | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93439 | &#39;T&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93457 | &#39;C&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 94008 | &#39;T&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 94035 | &#39;C&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-------------+----------------------------------------------------+</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">b_conv</span><span class="p">)</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;Chr&#39;  | &#39;Pos&#39; | &#39;Ref&#39; | &#39;Nref&#39; | &#39;Der&#39; | &#39;Mut&#39; | &#39;GeneId&#39;   | &#39;GeneAlias&#39;     | &#39;GeneDescr&#39;                                              |</span>
<span class="go">+========+=======+=======+========+=======+=======+============+=================+==========================================================+</span>
<span class="go">| &#39;MAL1&#39; | 91099 | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10,RIF&#39; | &#39;rifin&#39;                                                  |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 91104 | &#39;A&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10,RIF&#39; | &#39;rifin&#39;                                                  |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93363 | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93382 | &#39;T&#39;   | &#39;G&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93384 | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93390 | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93439 | &#39;T&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93457 | &#39;C&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 94008 | &#39;T&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 94035 | &#39;C&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
</pre></div>
</div>
<p>Now the tables are ready for comparison.</p>
</div>
<div class="section" id="looking-for-missing-or-unexpected-rows">
<h2>Looking for missing or unexpected rows<a class="headerlink" href="#looking-for-missing-or-unexpected-rows" title="Permalink to this headline">¶</a></h2>
<p>Because both tables should contain the same list of genome locations,
they should have the same number of rows. Use <a class="reference internal" href="index.html#petl.rowcount" title="petl.rowcount"><tt class="xref py py-func docutils literal"><span class="pre">rowcount()</span></tt></a> to
compare:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">a_conv</span><span class="p">)</span>
<span class="go">103647</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">b_conv</span><span class="p">)</span>
<span class="go">103618</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">103647</span><span class="o">-</span><span class="mi">103618</span>
<span class="go">29</span>
</pre></div>
</div>
<p>There is some discrepancy. First investigate by comparing just the
genomic locations, defined by the &#8216;Chr&#8217; and &#8216;Pos&#8217; fields, using
<a class="reference internal" href="index.html#petl.complement" title="petl.complement"><tt class="xref py py-func docutils literal"><span class="pre">complement()</span></tt></a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">a_locs</span> <span class="o">=</span> <span class="n">cut</span><span class="p">(</span><span class="n">a_conv</span><span class="p">,</span> <span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b_locs</span> <span class="o">=</span> <span class="n">cut</span><span class="p">(</span><span class="n">b_conv</span><span class="p">,</span> <span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">locs_only_in_a</span> <span class="o">=</span> <span class="n">complement</span><span class="p">(</span><span class="n">a_locs</span><span class="p">,</span> <span class="n">b_locs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">locs_only_in_a</span><span class="p">)</span>
<span class="go">29</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">locs_only_in_a</span><span class="p">)</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;Chr&#39;   | &#39;Pos&#39;   |</span>
<span class="go">+=========+=========+</span>
<span class="go">| &#39;MAL1&#39;  | 216961  |</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;MAL10&#39; | 538210  |</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;MAL10&#39; | 548779  |</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;MAL10&#39; | 1432969 |</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;MAL11&#39; | 500289  |</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;MAL11&#39; | 1119809 |</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;MAL11&#39; | 1278859 |</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;MAL12&#39; | 51827   |</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;MAL13&#39; | 183727  |</span>
<span class="go">+---------+---------+</span>
<span class="go">| &#39;MAL13&#39; | 398404  |</span>
<span class="go">+---------+---------+</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">locs_only_in_b</span> <span class="o">=</span> <span class="n">complement</span><span class="p">(</span><span class="n">b_locs</span><span class="p">,</span> <span class="n">a_locs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">locs_only_in_b</span><span class="p">)</span>
<span class="go">0</span>
</pre></div>
</div>
<p>So it appears that 29 locations are missing from table <cite>b</cite>. Export
these missing locations to a CSV file using <a class="reference internal" href="index.html#petl.tocsv" title="petl.tocsv"><tt class="xref py py-func docutils literal"><span class="pre">tocsv()</span></tt></a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">tocsv</span><span class="p">(</span><span class="n">locs_only_in_a</span><span class="p">,</span> <span class="s">&#39;missing_locations.csv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>A shorthand for finding the difference between two tables is the
<a class="reference internal" href="index.html#petl.diff" title="petl.diff"><tt class="xref py py-func docutils literal"><span class="pre">diff()</span></tt></a> function:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">locs_only_in_b</span><span class="p">,</span> <span class="n">locs_only_in_a</span> <span class="o">=</span> <span class="n">diff</span><span class="p">(</span><span class="n">a_locs</span><span class="p">,</span> <span class="n">b_locs</span><span class="p">)</span>
</pre></div>
</div>
<p>An alternative method for finding rows in one table where some key
value is not present in another table is to use the <a class="reference internal" href="index.html#petl.antijoin" title="petl.antijoin"><tt class="xref py py-func docutils literal"><span class="pre">antijoin()</span></tt></a>
function:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">locs_only_in_a</span> <span class="o">=</span> <span class="n">antijoin</span><span class="p">(</span><span class="n">a_conv</span><span class="p">,</span> <span class="n">b_conv</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">locs_only_in_a</span><span class="p">)</span>
<span class="go">29</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">locs_only_in_a</span><span class="p">)</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;Chr&#39;   | &#39;Pos&#39;   | &#39;Ref&#39; | &#39;Nref&#39; | &#39;Der&#39; | &#39;Mut&#39; | &#39;GeneId&#39;     | &#39;GeneAlias&#39;             | &#39;GeneDescr&#39;                                    |</span>
<span class="go">+=========+=========+=======+========+=======+=======+==============+=========================+================================================+</span>
<span class="go">| &#39;MAL1&#39;  | 216961  | &#39;T&#39;   | &#39;C&#39;    | &#39;C&#39;   | &#39;S&#39;   | &#39;PFA0245w&#39;   | &#39;MAL1P1.40&#39;             | &#39;novel putative transporter, PfNPT&#39;            |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;MAL10&#39; | 538210  | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PF10_0133&#39;  | &#39;-&#39;                     | &#39;hypothetical protein&#39;                         |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;MAL10&#39; | 548779  | &#39;A&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PF10_0136&#39;  | &#39;-&#39;                     | &#39;Initiation factor 2 subunit family, putative&#39; |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;MAL10&#39; | 1432969 | &#39;T&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PF10_0355&#39;  | &#39;-&#39;                     | &#39;Erythrocyte membrane protein, putative&#39;       |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;MAL11&#39; | 500289  | &#39;C&#39;   | &#39;A&#39;    | &#39;A&#39;   | &#39;N&#39;   | &#39;PF11_0528&#39;  | &#39;-&#39;                     | &#39;hypothetical protein&#39;                         |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;MAL11&#39; | 1119809 | &#39;G&#39;   | &#39;C&#39;    | &#39;C&#39;   | &#39;N&#39;   | &#39;PF11_0300&#39;  | &#39;-&#39;                     | &#39;hypothetical protein&#39;                         |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;MAL11&#39; | 1278859 | &#39;A&#39;   | &#39;G&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PF11_0341&#39;  | &#39;-&#39;                     | &#39;hypothetical protein&#39;                         |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;MAL12&#39; | 51827   | &#39;T&#39;   | &#39;G&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFL0030c&#39;   | &#39;2277.t00006,MAL12P1.6&#39; | &#39;erythrocyte membrane protein 1 (PfEMP1)&#39;      |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;MAL13&#39; | 183727  | &#39;G&#39;   | &#39;C&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;MAL13P1.18&#39; | &#39;-&#39;                     | &#39;hypothetical protein, conserved&#39;              |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
<span class="go">| &#39;MAL13&#39; | 398404  | &#39;G&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;MAL13P1.41&#39; | &#39;-&#39;                     | &#39;hypothetical protein, conserved&#39;              |</span>
<span class="go">+---------+---------+-------+--------+-------+-------+--------------+-------------------------+------------------------------------------------+</span>
</pre></div>
</div>
</div>
<div class="section" id="finding-conflicts">
<h2>Finding conflicts<a class="headerlink" href="#finding-conflicts" title="Permalink to this headline">¶</a></h2>
<p>We&#8217;d also like to compare the values given in the other fields, to
find any discrepancies between the two tables.</p>
<p>First, examine all fields for discrepancies, using the <a class="reference internal" href="index.html#petl.cat" title="petl.cat"><tt class="xref py py-func docutils literal"><span class="pre">cat()</span></tt></a> and
<a class="reference internal" href="index.html#petl.conflicts" title="petl.conflicts"><tt class="xref py py-func docutils literal"><span class="pre">conflicts()</span></tt></a> functions:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ab</span> <span class="o">=</span> <span class="n">cat</span><span class="p">(</span><span class="n">a_conv</span><span class="p">,</span> <span class="n">b_conv</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_conflicts</span> <span class="o">=</span> <span class="n">conflicts</span><span class="p">(</span><span class="n">ab</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">ab_conflicts</span><span class="p">)</span>
<span class="go">185978</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">ab_conflicts</span><span class="p">)</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;Chr&#39;  | &#39;Pos&#39; | &#39;Ref&#39; | &#39;Nref&#39; | &#39;Der&#39; | &#39;Mut&#39; | &#39;GeneId&#39;   | &#39;GeneAlias&#39;     | &#39;GeneDescr&#39;                                              |</span>
<span class="go">+========+=======+=======+========+=======+=======+============+=================+==========================================================+</span>
<span class="go">| &#39;MAL1&#39; | 91099 | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10&#39;     | &#39;rifin&#39;                                                  |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 91099 | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;S&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10,RIF&#39; | &#39;rifin&#39;                                                  |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 91104 | &#39;A&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10&#39;     | &#39;rifin&#39;                                                  |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 91104 | &#39;A&#39;   | &#39;T&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0095c&#39; | &#39;MAL1P1.10,RIF&#39; | &#39;rifin&#39;                                                  |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93363 | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93363 | &#39;T&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;hypothetical protein, conserved in P. falciparum&#39;       |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93382 | &#39;T&#39;   | &#39;G&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93382 | &#39;T&#39;   | &#39;G&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;hypothetical protein, conserved in P. falciparum&#39;       |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93384 | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93384 | &#39;G&#39;   | &#39;A&#39;    | &#39;-&#39;   | &#39;N&#39;   | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39;     | &#39;hypothetical protein, conserved in P. falciparum&#39;       |</span>
<span class="go">+--------+-------+-------+--------+-------+-------+------------+-----------------+----------------------------------------------------------+</span>
</pre></div>
</div>
<p>From a glance at the conflicts above, it appears there are
discrepancies in the &#8216;GeneAlias&#8217; and &#8216;GeneDescr&#8217; fields. There may
also be conflicts in other fields, so we need to investigate further.</p>
<p>Find conflicts, one field at a time, starting with the &#8216;Ref&#8217;, &#8216;Nref&#8217;,
&#8216;Der&#8217; and &#8216;Mut&#8217; fields:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ab_ref</span> <span class="o">=</span> <span class="n">cut</span><span class="p">(</span><span class="n">ab</span><span class="p">,</span> <span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">,</span> <span class="s">&#39;Ref&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_ref_conflicts</span> <span class="o">=</span> <span class="n">conflicts</span><span class="p">(</span><span class="n">ab_ref</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">ab_ref_conflicts</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_nref</span> <span class="o">=</span> <span class="n">cut</span><span class="p">(</span><span class="n">ab</span><span class="p">,</span> <span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">,</span> <span class="s">&#39;Nref&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_nref_conflicts</span> <span class="o">=</span> <span class="n">conflicts</span><span class="p">(</span><span class="n">ab_nref</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">ab_nref_conflicts</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_der</span> <span class="o">=</span> <span class="n">cut</span><span class="p">(</span><span class="n">ab</span><span class="p">,</span> <span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">,</span> <span class="s">&#39;Der&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_der_conflicts</span> <span class="o">=</span> <span class="n">conflicts</span><span class="p">(</span><span class="n">ab_der</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">ab_der_conflicts</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_mut</span> <span class="o">=</span> <span class="n">cut</span><span class="p">(</span><span class="n">ab</span><span class="p">,</span> <span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">,</span> <span class="s">&#39;Mut&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_mut_conflicts</span> <span class="o">=</span> <span class="n">conflicts</span><span class="p">(</span><span class="n">ab_mut</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">ab_mut_conflicts</span><span class="p">)</span>
<span class="go">3592</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">ab_mut_conflicts</span><span class="p">)</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;Chr&#39;  | &#39;Pos&#39;  | &#39;Mut&#39; |</span>
<span class="go">+========+========+=======+</span>
<span class="go">| &#39;MAL1&#39; | 99099  | &#39;-&#39;   |</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;MAL1&#39; | 99099  | &#39;N&#39;   |</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;MAL1&#39; | 99211  | &#39;-&#39;   |</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;MAL1&#39; | 99211  | &#39;N&#39;   |</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;MAL1&#39; | 197903 | &#39;N&#39;   |</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;MAL1&#39; | 197903 | &#39;S&#39;   |</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;MAL1&#39; | 384429 | &#39;N&#39;   |</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;MAL1&#39; | 384429 | &#39;S&#39;   |</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;MAL1&#39; | 513268 | &#39;N&#39;   |</span>
<span class="go">+--------+--------+-------+</span>
<span class="go">| &#39;MAL1&#39; | 513268 | &#39;S&#39;   |</span>
<span class="go">+--------+--------+-------+</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">tocsv</span><span class="p">(</span><span class="n">ab_mut_conflicts</span><span class="p">,</span> <span class="s">&#39;mut_conflicts.csv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>So there are some conflicts in the &#8216;Mut&#8217; field, and we&#8217;ve saved them to
a file for later reference.</p>
<p>We&#8217;d like also to find conflicts in the &#8216;GeneId&#8217;, &#8216;GeneAlias&#8217; and
&#8216;GeneDescr&#8217; fields. Here, while we&#8217;re focusing on finding conflicts in
one field at a time, we&#8217;ll include information from the other fields
in the output, to provide extra diagnostic information when
investigating the cause of any conflicts, making use of the <cite>include</cite>
keyword argument to the <a class="reference internal" href="index.html#petl.conflicts" title="petl.conflicts"><tt class="xref py py-func docutils literal"><span class="pre">conflicts()</span></tt></a> function:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">ab_gene</span> <span class="o">=</span> <span class="n">cut</span><span class="p">(</span><span class="n">ab</span><span class="p">,</span> <span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">,</span> <span class="s">&#39;GeneId&#39;</span><span class="p">,</span> <span class="s">&#39;GeneAlias&#39;</span><span class="p">,</span> <span class="s">&#39;GeneDescr&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_geneid_conflicts</span> <span class="o">=</span> <span class="n">conflicts</span><span class="p">(</span><span class="n">ab_gene</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">),</span> <span class="n">include</span><span class="o">=</span><span class="s">&#39;GeneId&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">ab_geneid_conflicts</span><span class="p">)</span>
<span class="go">5434</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">ab_geneid_conflicts</span><span class="p">)</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;Chr&#39;  | &#39;Pos&#39;  | &#39;GeneId&#39;   | &#39;GeneAlias&#39;           | &#39;GeneDescr&#39;                                       |</span>
<span class="go">+========+========+============+=======================+===================================================+</span>
<span class="go">| &#39;MAL1&#39; | 190628 | &#39;PFA0215w&#39; | &#39;MAL1P1.70&#39;           | &#39;hypothetical protein, conserved&#39;                 |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 190628 | &#39;PFA0220w&#39; | &#39;PFA0215w,MAL1P1.34b&#39; | &#39;ubiquitin carboxyl-terminal hydrolase, putative&#39; |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 190668 | &#39;PFA0215w&#39; | &#39;MAL1P1.70&#39;           | &#39;hypothetical protein, conserved&#39;                 |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 190668 | &#39;PFA0220w&#39; | &#39;PFA0215w,MAL1P1.34b&#39; | &#39;ubiquitin carboxyl-terminal hydrolase, putative&#39; |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 190786 | &#39;PFA0215w&#39; | &#39;MAL1P1.70&#39;           | &#39;hypothetical protein, conserved&#39;                 |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 190786 | &#39;PFA0220w&#39; | &#39;PFA0215w,MAL1P1.34b&#39; | &#39;ubiquitin carboxyl-terminal hydrolase, putative&#39; |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 190808 | &#39;PFA0215w&#39; | &#39;MAL1P1.70&#39;           | &#39;hypothetical protein, conserved&#39;                 |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 190808 | &#39;PFA0220w&#39; | &#39;PFA0215w,MAL1P1.34b&#39; | &#39;ubiquitin carboxyl-terminal hydrolase, putative&#39; |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 190854 | &#39;PFA0215w&#39; | &#39;MAL1P1.70&#39;           | &#39;hypothetical protein, conserved&#39;                 |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 190854 | &#39;PFA0220w&#39; | &#39;PFA0215w,MAL1P1.34b&#39; | &#39;ubiquitin carboxyl-terminal hydrolase, putative&#39; |</span>
<span class="go">+--------+--------+------------+-----------------------+---------------------------------------------------+</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">tocsv</span><span class="p">(</span><span class="n">ab_geneid_conflicts</span><span class="p">,</span> <span class="s">&#39;geneid_conflicts.csv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_genealias_conflicts</span> <span class="o">=</span> <span class="n">conflicts</span><span class="p">(</span><span class="n">ab_gene</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">),</span> <span class="n">include</span><span class="o">=</span><span class="s">&#39;GeneAlias&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">ab_genealias_conflicts</span><span class="p">)</span>
<span class="go">39144</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">ab_genealias_conflicts</span><span class="p">)</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;Chr&#39;  | &#39;Pos&#39;  | &#39;GeneId&#39;   | &#39;GeneAlias&#39;                              | &#39;GeneDescr&#39;                                 |</span>
<span class="go">+========+========+============+==========================================+=============================================+</span>
<span class="go">| &#39;MAL1&#39; | 91099  | &#39;PFA0095c&#39; | &#39;MAL1P1.10&#39;                              | &#39;rifin&#39;                                     |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 91099  | &#39;PFA0095c&#39; | &#39;MAL1P1.10,RIF&#39;                          | &#39;rifin&#39;                                     |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 91104  | &#39;PFA0095c&#39; | &#39;MAL1P1.10&#39;                              | &#39;rifin&#39;                                     |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 91104  | &#39;PFA0095c&#39; | &#39;MAL1P1.10,RIF&#39;                          | &#39;rifin&#39;                                     |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 99099  | &#39;PFA0110w&#39; | &#39;MAL1P1.13,Pf155&#39;                        | &#39;ring-infected erythrocyte surface antigen&#39; |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 99099  | &#39;PFA0110w&#39; | &#39;MAL1P1.13,Pf155,RESA&#39;                   | &#39;ring-infected erythrocyte surface antigen&#39; |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 99211  | &#39;PFA0110w&#39; | &#39;MAL1P1.13,Pf155&#39;                        | &#39;ring-infected erythrocyte surface antigen&#39; |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 99211  | &#39;PFA0110w&#39; | &#39;MAL1P1.13,Pf155,RESA&#39;                   | &#39;ring-infected erythrocyte surface antigen&#39; |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 111583 | &#39;PFA0125c&#39; | &#39;JESEBL,jesebl,eba-181,EBA181,MAL1P1.16&#39; | &#39;erythrocyte binding antigen-181&#39;           |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 111583 | &#39;PFA0125c&#39; | &#39;jesebl,JESEBL,eba-181,MAL1P1.16&#39;        | &#39;erythrocyte binding antigen-181&#39;           |</span>
<span class="go">+--------+--------+------------+------------------------------------------+---------------------------------------------+</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">tocsv</span><span class="p">(</span><span class="n">ab_genealias_conflicts</span><span class="p">,</span> <span class="s">&#39;genealias_conflicts.csv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab_genedescr_conflicts</span> <span class="o">=</span> <span class="n">conflicts</span><span class="p">(</span><span class="n">ab_gene</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;Chr&#39;</span><span class="p">,</span> <span class="s">&#39;Pos&#39;</span><span class="p">),</span> <span class="n">include</span><span class="o">=</span><span class="s">&#39;GeneDescr&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rowcount</span><span class="p">(</span><span class="n">ab_genedescr_conflicts</span><span class="p">)</span>
<span class="go">177730</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">look</span><span class="p">(</span><span class="n">ab_genedescr_conflicts</span><span class="p">)</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;Chr&#39;  | &#39;Pos&#39; | &#39;GeneId&#39;   | &#39;GeneAlias&#39; | &#39;GeneDescr&#39;                                              |</span>
<span class="go">+========+=======+============+=============+==========================================================+</span>
<span class="go">| &#39;MAL1&#39; | 93363 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93363 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39;       |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93382 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93382 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39;       |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93384 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93384 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39;       |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93390 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93390 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39;       |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93439 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;Plasmodium exported protein (PHISTa), unknown function&#39; |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>
<span class="go">| &#39;MAL1&#39; | 93439 | &#39;PFA0100c&#39; | &#39;MAL1P1.11&#39; | &#39;hypothetical protein, conserved in P. falciparum&#39;       |</span>
<span class="go">+--------+-------+------------+-------------+----------------------------------------------------------+</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">tocsv</span><span class="p">(</span><span class="n">ab_genedescr_conflicts</span><span class="p">,</span> <span class="s">&#39;genedescr_conflicts.csv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>So each of these three fields has a different set of conflicts, and
we&#8217;ve saved the output to CSV files for later reference.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Case Study 1 - Comparing Tables</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#preparing-the-data">Preparing the data</a></li>
<li><a class="reference internal" href="#looking-for-missing-or-unexpected-rows">Looking for missing or unexpected rows</a></li>
<li><a class="reference internal" href="#finding-conflicts">Finding conflicts</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">petl - Extract, Transform and Load (Tables of Data)</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="related_work.html"
                        title="next chapter">Related Work</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/case_study_1.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="related_work.html" title="Related Work"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="petl - Extract, Transform and Load (Tables of Data)"
             >previous</a> |</li>
        <li><a href="index.html">petl v0.8 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, Alistair Miles.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.7.
    </div>
  </body>
</html>